bundles/remote_services/discovery_zeroconf/src/discovery_zeroconf_announcer.c

```diff
@@ -0,0 +1,533 @@
+eventfd_err:
+    free(announcer);
+    return status;
+}
+
+void discoveryZeroconfAnnouncer_destroy(discovery_zeroconf_announcer_t *announcer) {
+    celixThreadMutex_lock(&announcer->mutex);
+    announcer->running= false;
+    celixThreadMutex_unlock(&announcer->mutex);
+    discoveryZeroconfAnnouncer_eventNotify(announcer);
+    celixThread_join(announcer->refreshEPThread, NULL);
+    celix_bundleContext_unregisterServiceAsync(announcer->ctx, announcer->epListenerSvcId, NULL, NULL);
+    celix_bundleContext_waitForAsyncUnregistration(announcer->ctx, announcer->epListenerSvcId);
+
+    announce_endpoint_entry_t *entry = NULL;
+    int size = celix_arrayList_size(announcer->revokedEndpoints);
+    for (int i = 0; i < size; ++i) {
+        entry = (announce_endpoint_entry_t *)celix_arrayList_get(announcer->revokedEndpoints, i);
+        celix_properties_destroy(entry->properties);
+        free(entry);
+    }
+    celix_arrayList_destroy(announcer->revokedEndpoints);
+
+    CELIX_STRING_HASH_MAP_ITERATE(announcer->endpoints,iter) {
+        entry = (announce_endpoint_entry_t *) iter.value.ptrValue;
+        celix_properties_destroy(entry->properties);
+        free(entry);
+    }
+    celix_stringHashMap_destroy(announcer->endpoints);
+
+    celixThreadMutex_destroy(&announcer->mutex);
+    close(announcer->eventFd);
+    free(announcer);
+    return;
+}
+
+static void discoveryZeroconfAnnouncer_eventNotify(discovery_zeroconf_announcer_t *announcer) {
+    eventfd_t val = 1;
+    eventfd_write(announcer->eventFd, val);
+    return;
+}
+
+static bool isLoopBackNetInterface(int ifIndex) {
+    if (ifIndex <= 0) {
+        return false;
+    }
+    bool loopBack = false;
+    int fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (fd >= 0) {
+        struct ifreq ifr;
+        memset(&ifr, 0, sizeof(ifr));
+        if (if_indextoname((unsigned int)ifIndex, ifr.ifr_name) != NULL) {
+            if (ioctl(fd, SIOCGIFFLAGS, &ifr) == 0) {
+                loopBack = !!(ifr.ifr_ifru.ifru_flags & IFF_LOOPBACK);
+            }
+        }
+        close(fd);
+    }
+    return loopBack;
+}
+
+static  celix_status_t discoveryZeroconfAnnouncer_endpointAdded(void *handle, endpoint_description_t *endpoint, char *matchedFilter) {
+    (void)matchedFilter;//unused
+    celix_status_t status = CELIX_SUCCESS;
+    discovery_zeroconf_announcer_t *announcer = (discovery_zeroconf_announcer_t *)handle;
+    assert(announcer != NULL);
+    if (endpointDescription_isInvalid(endpoint)) {
+        celix_logHelper_error(announcer->logHelper, "Announcer: Endpoint is invalid.");
+        return CELIX_ILLEGAL_ARGUMENT;
+    }
+
+    celix_logHelper_info(announcer->logHelper, "Announcer: Add endpoint for %s(%s).", endpoint->serviceName, endpoint->id);
+
+    announce_endpoint_entry_t *entry = calloc(1, sizeof(*entry));
+    assert(entry != NULL);
+    entry->registerRef = NULL;
+    entry->announced = false;
+    entry->uid = celix_utils_stringHash(endpoint->id);
```

Comment:
hash should not be needed. the endpoint id should be unique and I think it is always a UUID.

Reply:
To avoid mDNS service instance name conflicts on localonly interface, in the code, the instance name consists of the service name and the UID. But the maximum size of an mDNS instance name is 64 bytes, it is not enough to hold the service name and  "endpoint->id". Therefore, we use the hash of "endpoint->id".

We can also just set the  "endpoint->id" to the mDNS service instance, but it may affect the readability of the mDNS service instance.

Reply:
If the usage of mDNS service instance is possible I think this is better, because a hash is not guaranteed to be unique.